woocommerce/woocommerce-ios
# Enforce use of bundle exec in Fastfile+Podfile

## Description

This adds checks in the `Fastfile` and `Podfile` so that it stops you from accidentally run `fastlane` or `pod` without prefixing them with `bundle exec`.

If you forget to prepend `bundle exec` when invoking `fastlane` or `pod`, this will now error and stop, telling you to use `bundle exec` – so you run the versions of those tools specified in the repo's `Gemfile.lock` – instead of continuing to run with the system-wide version – which might not be the same and behave differently than when ran under bundle exec.

## To test:

1️⃣  Run `pod install` and see it fail.

<details><summary>Expected pod failure when running without <tt>bundle exec</tt></summary>

```
[!] Invalid `Podfile` file: Please run CocoaPods via `bundle exec`.

 #  from /Users/olivier/Documents/Dev/woocommerce-ios/Podfile:4
 #  -------------------------------------------
 #  unless ['BUNDLE_BIN_PATH', 'BUNDLE_GEMFILE'].any? { |k| ENV.key?(k) }
 >    raise 'Please run CocoaPods via `bundle exec`'
 #  end
 #  -------------------------------------------
```

(Note: Sadly there is no way to avoid the stacktrace from showing, which might make you think there is a syntax error in the Podfile while the syntax is correct and it just shows where the exception was raised)
</details>

2️⃣ Run `bundle exec pod install` (or `be pod install` if you created an alias for `bundle exec` in your `.zshrc`, which I recommend) and check that it runs as expected

3️⃣ Run `fastlane lanes`, and check it fails with the following output

<details><summary>Expected fastlane failure when running without <tt>bundle exec</tt></summary>

```
[✔] 🚀 
[17:58:16]: fastlane detected a Gemfile in the current directory
[17:58:16]: However, it seems like you didn't use `bundle exec`
[17:58:16]: To launch fastlane faster, please use
[17:58:16]: 
[17:58:16]: $ bundle exec fastlane lanes
[17:58:16]: 
[17:58:16]: Get started using a Gemfile for fastlane https://docs.fastlane.tools/getting-started/ios/setup/#use-a-gemfile
[17:58:17]: Installing Ruby gem 'fastlane-plugin-wpmreleasetoolkit'...
[17:58:33]: Found gem "fastlane-plugin-release_helper" instead of the required name "fastlane-plugin-wpmreleasetoolkit"
[17:58:33]: Successfully installed 'fastlane-plugin-wpmreleasetoolkit'
[17:58:33]: Error loading plugin 'fastlane-plugin-wpmreleasetoolkit': cannot load such file -- fastlane/plugin/wpmreleasetoolkit
+-----------------------------------+-----------+---------------------------------------------------+
|                                           Used plugins                                            |
+-----------------------------------+-----------+---------------------------------------------------+
| Plugin                            | Version   | Action                                            |
+-----------------------------------+-----------+---------------------------------------------------+
| fastlane-plugin-wpmreleasetoolkit | undefined | No actions found                                  |
| fastlane-plugin-sentry            | 1.6.0     | sentry_upload_dsym sentry_set_commits             |
|                                   |           | sentry_upload_sourcemap sentry_finalize_release   |
|                                   |           | sentry_upload_file sentry_upload_proguard         |
|                                   |           | sentry_create_release                             |
| fastlane-plugin-appcenter         | 1.8.0     | appcenter_fetch_version_number                    |
|                                   |           | appcenter_fetch_devices appcenter_upload          |
| fastlane-plugin-test_center       | 3.14.2    | suppressed_tests suppress_tests                   |
|                                   |           | tests_from_xcresult suppress_tests_from_junit     |
|                                   |           | tests_from_xctestrun quit_core_simulator_service  |
|                                   |           | collate_test_result_bundles tests_from_junit      |
|                                   |           | multi_scan test_options_from_testplan             |
|                                   |           | collate_html_reports collate_junit_reports        |
|                                   |           | collate_json_reports testplans_from_scheme        |
|                                   |           | collate_xcresults                                 |
+-----------------------------------+-----------+---------------------------------------------------+

[!] No actions were found while loading one or more plugins
    Please use `bundle exec fastlane` with plugins
    More info - https://docs.fastlane.tools/plugins/using-plugins/#run-with-plugins

[17:58:34]: ------------------------------
[17:58:34]: --- Step: default_platform ---
[17:58:34]: ------------------------------

[!] Please run fastlane via `bundle exec`
```
</details>

4️⃣ Run `bundle exec fastlane lanes` and check that it runs without issues and lists the available lanes
